### PR TITLE
make dragging links do more

### PR DESCRIPTION
**Pull request by BurningTreeC**
one can just press the ctrl key while dragging so that dragged links create transclusions
normal: [[tiddler title]]
with ctrl: {{tiddler title}}
with shift: **your prefix**tiddler title**your suffix**

This works for me, though I don't know if it breaks something somewhere else
Simon

### CODE REVIEW

**Comment by Jermolene** [comment]
Hi @BurningTreeC this is a cracking idea, and the implementation looks good. Can you please update the end user documentation? The existing "Drag and Drop" tiddler is quite long, so perhaps we could add a new tiddler "Drag and Drop Modifier Keys", and link to it from a single sentence in the former?

**Comment by BurningTreeC** [comment]
thanks @Jermolene,
I've added the Modifier tiddler and referenced it
hope it's ok

**Comment by BurningTreeC** [comment]
now with this implementation Ctrl+Shift+drag could also be enabled to be configurable,
I'm just without ideas
If I ping the group about this and something good comes out, could we add it?

**Comment by pmario** [comment]
Having a look at this topic, I think it should be a plugin!

https://groups.google.com/forum/#!topic/tiddlywiki/aUJNCCnR26Q

I personally prefer a configuration option in this case. 

Would it be possible to implement it like these https://tiddlywiki.com/#KeyboardShortcuts:KeyboardShortcuts%20%24%3A%2FControlPanel settings.

**Comment by BurningTreeC** [comment]
Hi @pmario, please have another look at the topic

the Shift+drag = prefix + title + suffix is already in this pr
the Idea-gathering is for figuring out if there's a very good Idea that can be discussed, otherwise I wouldn't touch it further

Simon

**Comment by pmario** [comment]
As I wrote in the group. ... I support this PR, because I think it is generic enough for the core. 

But with more possibilities I would prefer to have it configurable, similar to the existing KeyStroke settings. .. if possible. and if can keep the new UI small ;)

**Comment by BurningTreeC** [comment]
Thank you @pmario!
IF more possibilities, how about adding a small UI below the current keyboard shortcuts UI?
The options wouldn't be many, it's somehow related to the editor keyboard shortcuts, adding a new tab somewhere else isn't necessary I think

/IF

**Comment by BurningTreeC** [comment]
Hi again, there are exactly zero ideas,
I think I don't want to touch it further ;)

**Comment by BurningTreeC** [comment]
I've added two lines to have it work consistently
Only titles with spaces get [[ as prefix and ]] as suffix
This leads to some kind of inconsistency when dragging links normally, without modifiers

Three tiddlers:
1) I have spaces     --> [[I have spaces]] ... and I'm a link
2) ImACamel      -->  ImACamel     ... and a link
3) imlowercaseonly    --> imlowercaseonly ... without [[    ]]

my question is, if this is needed/wanted or if I should make the [[ ]] default for all titles one of these:

1) titles starting with lowercase letters and not already having [[   ]]  get [[  ]]  - titles starting with uppercase get checked if theres another uppercase, checking for spaces not needed, then they won't get [[  ]]

2) they all get [[   ]]

**Comment by BurningTreeC** [comment]
Hi all, I've given this some thoughts:

the default dragging behaviour is to drag just the title string, but there are many tiddlers being dragged that contain spaces, so that links get created because [[  ]] get added automatically

I would let the ctrl key create transclusions, the **shift key create links**  and for prefix and suffix use ctrl + shift 

so that for normal dragging without modifier keys the titlestring can be checked for [[  ]] and they can be removed

that would be a choice pro consistency

**Comment by pmario** [comment]
At the moment I think there is a problem. 

Dragging links creates `[[link with spaces]]` and `linkwithoutspaces` ... that's wrong
imo it should be: `[[link with spaces]]` and `[[linkwithoutspaces]]` ... no modifier keys needed.

So dragging a link, should create a link ... no extra key needed. 
Dragging text should create text ... no extra key needed

-----------

CTRL drag ... should create transclusions ... always
SHIFT ... my use prefix / postfix modifyers ... always

--------

Dragging CamelCase links should create CamelCase text since there is a core option to switch off/on CamelCase link rendering. We shouldn't mess with it. so _no_ `[[CamelCase]]` with drag and drop.

**Comment by BurningTreeC** [comment]
@pmario 
because of the core option to switch CamelCase on/off there would be checks needed for 
a) is it CamelCase
b) is the option set or unset

TW also renders system tiddlers as links without having to add brackets
TW also renders http://whatever.co.uk as links

So I'd have to distinguish between all those options

<hr>

At the end with what you propose there's no option to have just the title string itself
It's strange for example when I drag a tiddler link into a search field - having the brackets there, when I just want the title

I somehow care about the plain title string

the title string carer

**Comment by pmario** [comment]
>It's strange for example when I drag a tiddler link into a search field - having the brackets there, when I just want the title

I didn't think about this.

>I somehow care about the plain title string

I see ;)

File: core/modules/utils/dom/dragndrop.js [hunk 1]
```diff
@@ -54,17 +54,28 @@ exports.makeDraggable = function(options) {
...
-                    		var dragShiftTiddler = $tw.wiki.getTiddler("$:/config/DragLinkShift");
-                		if(dragShiftTiddler.fields["prefix"] !== undefined && dragShiftTiddler.fields["suffix"] !== undefined) {
-                        		var titleStringPrefix = dragShiftTiddler.fields["prefix"];
-                        		var titleStringSuffix = dragShiftTiddler.fields["suffix"];
-                        		if(/^\$:\/.*/.test(titleString)) {
-                                		titleString = titleStringPrefix + titleString + titleStringSuffix;
-                        		} else {
-                                		titleString = titleStringPrefix + titleString.replace(/\[/g, '').replace(/]/g, '') + titleStringSuffix;
-                        		}
+			  var dragShiftTiddler = $tw.wiki.getTiddler("$:/config/DragLinkShift");
```
Comment (pmario) [inline core/modules/utils/dom/dragndrop.js]: What happens if the config tiddler doesn't exist? ... Does it throw an undefined error?

File: core/modules/utils/dom/dragndrop.js [hunk 2]
```diff
@@ -54,17 +54,28 @@ exports.makeDraggable = function(options) {
...
+			  if(dragShiftTiddler.fields["prefix"] !== undefined && dragShiftTiddler.fields["suffix"] !== undefined) {
```
Comment (pmario) [inline core/modules/utils/dom/dragndrop.js]: imo it will throw an error here if `dragShiftTiddler` is undefined

**Comment by pmario** [comment]
I'm sorry, but I think, we created a bit of a mess here. ... The original code was already "spaghetti like" and the new nested ifs and additional checks imo make it hard to maintain now. ... 

Especially tests like this one: 

```
		var isUrl = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,4}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/.test(titleString); // see https://stackoverflow.com/questions/3809401/what-is-a-good-regular-expression-to-match-a-url
```

imo should be functions of $tw.utils ... or we should have a look how the core does it, to be consistent with the parser rules. ... There is some code somewhere, that does define regexp global tests like this one. ... but I can't find it atm. 

I would propose, to go back to the drawing boards and define a consistent functionality we want to have for all drag and drop actions. ... and then rewrite it in a consistent way.

File: core/modules/utils/dom/dragndrop.js (hunk 1 above)
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: no you're right it does

File: core/modules/utils/dom/dragndrop.js (hunk 2 above)
Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: yes you're totally right

Comment (BurningTreeC) [reply core/modules/utils/dom/dragndrop.js]: sorry for my indenting mess, I've mixed editors here

**Comment by BurningTreeC** [comment]
Ok @pmario, I've seen your inline comments...

I would close this pr in favour of a remake, using core-utils where there are

A decision about having one combination of dragging+modifiers that creates the plain titles only is still outstanding

Can we make this clear so that before starting I know how the final thing should behave?

**Comment by BurningTreeC** [comment]
For the check if the string starts with http/https... I've found **$tw.utils.isLinkExternal(link)** in
$:/core/modules/utils/utils.js

For checking for camelcase I probably should use **$tw.config.textPrimitives** from 
$:/core/modules/config.js

**Pull request closed without merging**